src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+      val bytes = barcodeBytes.clone()
+      SequenceUtil.reverseComplement(bytes)
+      bytes
+    }
+  }
+
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call. Observed
+    * qualities may not be null. */
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
+    var idx = 0
+    var count = 0
+    while (idx < observedBases.length) {
+      val expectedBase = expectedBases(idx)
+      val observedBase = observedBases(idx)
+      if (!SequenceUtil.isNoCall(expectedBase) && !SequenceUtil.basesEqual(observedBase, expectedBase)) {
+        count += 1
+      }
+      idx += 1
+    }
+    count
+  }
+
+  /** Reads multiple FASTQs at the same time, ensuring their read names match along the way.  This is
+    * useful for when we don't have a single interleaved FASTQ. */
+  private[fastq] class ParallelFastqReader(fastq: Seq[Path], fastqs: Seq[Path]*) extends Iterator[Seq[FastqRecord]] {
+
+    val readers: Seq[SequentialFastqSource] = (fastq +: fastqs.toSeq).map { otherFastqs: Seq[Path] =>
+      if (otherFastqs.size != fastq.size) throw new IllegalArgumentException("List of list fastqs had differing lengths.")
+      SequentialFastqSource(otherFastqs)
+    }
+
+    def next: Seq[FastqRecord] = {
+      val records = readers.map(_.next())
+      verifySameReadNames(records)
+      records
+    }
+
+    def hasNext: Boolean = {
+      if (readersHaveNext(true)) true
+      else if (readersHaveNext(false)) false
+      else throw new IllegalArgumentException("Fastqs have differing lengths")
+    }
+
+    private def readersHaveNext(desiredValue: Boolean): Boolean = readers.forall(_.hasNext == desiredValue)
+
+    private def verifySameReadNames(records: Seq[FastqRecord]): Unit = {
+      val readName = records.head.name
+      if (records.exists(_.name != readName)) {
+        throw new IllegalArgumentException(String.format("Mismatching read names in FASTQS:\n%s\n%s\n", readName, records.find(_.name != readName).get))
+      }
+    }
+
+    def close() = readers.foreach(_.close())
+  }
+
+  /** Helper case class to store the result of converting SAM records. */
+  private case class AssignedRecords(sampleOrdinal: Int, records: Seq[SAMRecord])
+
+  /** Helper class to convert fastq records into SAM records and write them out.
+    *
+    * @param input the iterator through groups of FASTQ records (r1, r2, i7, and i5).
+    * @param headers the headers, one per sample
+    * @param sampleSheet the sample sheet describing all the samples
+    * @param readStructures the read structures info, one for each expected fastq record that will be passed
+    *                           into `convertFastqsAndWriteRecords`.
+    */
+  private class FastqConverter(val input: Iterator[Seq[FastqRecord]],
```

Comment:
One thing I'd like to do in the near future is write a replacement for Picard's FastqToSam which has become overly complicated and doesn't support on-the-fly extraction of UMIs.  How hard would it be to refactor this class so that we had one per sample instead of one per sample sheet?  Given that it would be fairly trivial to write the single sample FastqToSam replacement.